### PR TITLE
Refactor event typing

### PR DESCRIPTION
**Pull request by remmycat**

### CODE REVIEW

**Review by ChristianHoffmannS2 — APPROVED** [review]

File: packages/ui-library/src/components/icon/index.ts [hunk 1]
```diff
@@ -9,7 +9,15 @@ import { until } from 'lit-html/directives/until.js';
...
 import { TAG_NAME } from './renderFunction';
 import { ThemeType } from '../../foundation/_tokens-generated/index.themes';
+import { BlrClickEvent, createBlrClickEvent } from '../../globals/events';
 
+export type BlrIconEventHandlers = {
+  blrClick?: (event: BlrClickEvent) => void;
+};
+
+/**
+ * @fires blrClick Icon got clicked
```
Comment (angsherpa456) [inline packages/ui-library/src/components/icon/index.ts:19]: For consistency, just like above "Icon was clicked" might be better

File: packages/ui-library/src/components/icon/index.ts [hunk 2]
```diff
@@ -20,13 +28,8 @@ export class BlrIcon extends LitElement {
...
 
-  // these are not triggered directly but allows us to map it internally and bve typesafe
-  @property() blrClick?: () => void;
-
-  protected handleClick = (event: Event) => {
-    this.dispatchEvent(
-      new CustomEvent('blrClick', { bubbles: true, composed: true, detail: { originalEvent: event } })
-    );
+  protected handleClick = (event: MouseEvent | KeyboardEvent) => {
+    this.dispatchEvent(createBlrClickEvent({ originalEvent: event }));
```
Comment (angsherpa456) [inline packages/ui-library/src/components/icon/index.ts:32]: Curly bracket can be removed

File: packages/ui-library/src/components/text-input/index.ts [hunk 3]
```diff
@@ -14,7 +14,30 @@ import { BlrFormCaptionRenderFunction } from '../form-caption/renderFunction';
...
+  blrBlur?: (event: BlrBlurEvent) => void;
+  blrTextValueChange?: (event: BlrTextValueChangeEvent) => void;
+  blrSelect?: (event: BlrSelectEvent) => void;
+};
+
+/**
+ * @fires blrFocus TextInput received focus
+ * @fires blrBlur TextInput lost focus
+ * @fires blrTextValueChange TextInput value changed
+ * @fires blrSelect Text in input got selected
```
Comment (angsherpa456) [inline packages/ui-library/src/components/text-input/index.ts:39]: Can you also change "blrSelect Text in input got selected" to "blrSelect Text in TextInput got selected" for consistency.

File: packages/ui-library/src/globals/events.ts [hunk 4]
```diff
@@ -0,0 +1,79 @@
+export type BlrFocusEventDetail = {
+  originalEvent: FocusEvent;
+};
+export type BlrFocusEvent = CustomEvent<BlrFocusEventDetail>;
+export const BlrFocusEventName = 'blrFocus';
+export function createBlrFocusEvent(detail: BlrFocusEventDetail): BlrFocusEvent {
```
Comment (angsherpa456) [inline packages/ui-library/src/globals/events.ts:6]: Nice approach 👍

File: packages/ui-library/src/components/icon/index.ts (hunk 2 above)
Comment (remmycat) [reply packages/ui-library/src/components/icon/index.ts:32]: It probably shouldn't be - `this.dispatchEvent()` returns a boolean, if I remove the brackets `handleClick` returns a boolean too, which doesn't make sense and could theoretically have an unexpected effect where handleClick is used.

Comment (angsherpa456) [reply packages/ui-library/src/components/icon/index.ts:32]: Aah, true, make sense, thanks for the insight.

**Review by angsherpa456 — DISMISSED** [review]
LGTM

**Review by ChristianHoffmannS2 — APPROVED** [review]

**Review by angsherpa456 — APPROVED** [review]
LGTM

**Pull request merged**